### PR TITLE
FreeBSD: Fix to build APPLEN

### PR DESCRIPTION
**Pull request by kiyolee**
Some tweaks to get APPLEN target to build on FreeBSD.
Not yet verify if the build actually works on FreeBSD.

### CODE REVIEW

File: source/CMakeLists.txt [hunk 1]
```diff
@@ -250,10 +250,22 @@ target_include_directories(appleii PRIVATE
...
 # this one appears in header files
+if ("${CMAKE_SYSTEM_NAME}" STREQUAL "FreeBSD")
+# FreeBSD could already have "zip.h" (from libzip) in /usr/local/include
+# which takes precedence than the one in /usr/local/include/minizip.
+# "zip.h" from libzip is not what we want here.
+# The following changes to use include path /usr/local/include for minizip
+# so that the code can include "minizip/zip.h" explicitly.
+target_include_directories(appleii PUBLIC
+  ${CMAKE_CURRENT_BINARY_DIR}  # for config.h
+  ${MINIZIP_INCLUDE_DIRS}/..
```
Comment (audetto) [inline source/CMakeLists.txt:261]: I see the issue, but it is not limited to FreeBSD. What happens if I install https://packages.ubuntu.com/noble/libzip-dev ?
There must be a more general solution

Comment (audetto) [reply source/CMakeLists.txt:261]: I have tried and there are no issues. The reason must be
1) if I want minizip, then I add `-I/usr/include/minizip' and although `zip.h` is ambiguous, this explicit `-I` wins over the default one (used by `libzip`)
2) If I want zip, no problem as I do not add the `minizip` `-I`

So, the issue is probably related to a user who have both `libzip` and `minizip` in `/usr/local` which breaks the (weak) order of includes.

Comment (audetto) [reply source/CMakeLists.txt:261]: OK, now I see better. In FreeBSD they are all installed in `/usr/local` so they cannot coexist.
Which is really odd, maybe nobody is using it, someone should file a bug.

The fact is that the command `pkg-config minizip -cflags` in FreeBSD returns `-I/usr/local/include/minizip'.

File: source/Tfe/DNS.cpp [hunk 2]
```diff
@@ -25,6 +25,9 @@ Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 #ifndef _MSC_VER
 #include <arpa/inet.h>
 #include <netdb.h>
+#ifdef __FreeBSD__
+#include <sys/socket.h> // AF_INET.
```
Comment (audetto) [inline source/Tfe/DNS.cpp:29]: no need to add `ifdef`, I will commit this.

File: source/linux/network/portfwds.cpp [hunk 3]
```diff
@@ -7,6 +7,9 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 #include <netdb.h>
+#ifdef __FreeBSD__
+#include <netinet/in.h> // sockaddr_in.
```
Comment (audetto) [inline source/linux/network/portfwds.cpp:11]: will commit

File: source/Uthernet2.cpp [hunk 4]
```diff
@@ -75,6 +75,9 @@ typedef int socklen_t;
 #include <fcntl.h>
 #include <sys/types.h>
 #include <errno.h>
+#ifdef __FreeBSD__
+#include <netinet/in.h> // sockaddr_in.
```
Comment (audetto) [inline source/Uthernet2.cpp:79]: no need of the `ifdef`, but this has to go to AppleWin

File: source/CMakeLists.txt (hunk 1 above)
Comment (kiyolee) [reply source/CMakeLists.txt:261]: I think it is exactly the order of `-I` problem. `pkg-config` is returning the right path for `minizip`. But some other libs is likely returning the common `/usr/local/include` and happens to be earlier in the include path. The kind of problem does happen to FreeBSD ports occasionally too. Ideally for FreeBSD `/usr/local/include` should always be the last path. But that's very difficult to control. Neither is to have `/usr/local/include/minizip` always in front, just not obvious to me how within CMake that can be done reliably. If all Linux distros always put `minizip` headers in its own subdirectory, the hack here could work universally and so in AppleWin code we could simply `#include <minizip/zip.h>`.

File: source/Tfe/DNS.cpp (hunk 2 above)
Comment (kiyolee) [reply source/Tfe/DNS.cpp:29]: Agree with that. This is just to explain the PR.
Should I update my PR? Or would you just fix your code in your way and drop the PR?
That's totally fine.

**Comment by kiyolee** [comment]
Note: APPLEN does run on FreeBSD, at least can boot DOS 3.3. The first observed problem is somehow only `F2` key works. None of the other listed function keys work.

**Comment by kiyolee** [comment]
FYI: APPLEN does run on FreeBSD, at least can boot DOS 3.3.

**Comment by audetto** [comment]
Hi

the `minizip` issue is odd. I have found an alternative solution.
It is not as stable as yours, but I can pretend it does not exist.

https://github.com/audetto/AppleWin/tree/freebsd2

I will try to upstream it to AppleWin, they should not care.

I have opened an issue https://github.com/madler/zlib/issues/977, feel free to chip in.

Then, as you say it works, but the ncurse F scan codes must be all over the place.

**Comment by audetto** [comment]
Actually it is Shift-F behaving oddly, so I added `F4` to exit.

**Comment by kiyolee** [comment]
An alternative solution is to drop minizip detection at all on FreeBSD. Very likely `/usr/local/include` would already be in the include path and `#include <minizip/zip.h>` might just work. That may work for Linux too as `/usr/include` is always there in the include path anyway.

**Comment by kiyolee** [comment]
> Actually it is Shift-F behaving oddly, so I added `F4` to exit.

I could only `Ctrl+\` (Quit signal) to exit which is not nice.

**Comment by audetto** [comment]
The most elegant solution would be to move `minizip` down 1 level inside AW, so that we all include it the same.

I see you have updated it recently, do you fancy proposing a new PR upstream?

**Comment by audetto** [comment]
minizip has removed the source of ambiguity.

https://github.com/madler/zlib/commit/7e6f0784cc0c33e8d5fcb368248168c6656f73c8

so, I will try to change AppleWin upstream for the new layout of minizip, otherwise I will just incorporate these changes. directly.

**Comment by audetto** [comment]
I think all these changes are now available in master. Could you please check.

**Comment by kiyolee** [comment]
Affirmative. Thanks.

**Pull request closed without merging**